### PR TITLE
Output unframed (garbage) data to file

### LINKED ISSUE #2378: GDS Logs in Triplicate
| | |
|:---|:---|
|**_F´ Version_**| |
|**_Affected Component_**|   |
---
## Problem Description

Log messages in GDS are printed thrice.

## How to Reproduce

Look at any Python GDS log (e.g. comm.py.log).

## Expected Behavior

Single log output.

### LINKED ISSUE #2379: GDS Log Scrolling
| | |
|:---|:---|
|**_F´ Version_**| |
|**_Affected Component_**|  |
---
## Feature Description

Add in GDS log pane scrolling as it will make life better when reviewing logs.

### LINKED ISSUE #2380: Stop Clearing Log Pane Output on Shutdown
| | |
|:---|:---|
|**_F´ Version_**| |
|**_Affected Component_**|   |
---
## Problem Description

When reading logs from the log pane in the GDS, the current log is cleared on shutdown and replaced with [ERROR].  This can prevent seeing errors and reviewing and does not add context.

### LINKED ISSUE #2381: Log Dropdown Jumps About
| | |
|:---|:---|
|**_F´ Version_**| |
|**_Affected Component_**|  |
---
## Feature Description

The log pane drop down jumps about once a second.  This makes selecting logs difficult.

### LINKED ISSUE #2382: Allow GDS To Output Unframed Data
| | |
|:---|:---|
|**_F´ Version_**| |
|**_Affected Component_**|  |
---
## Feature Description

The GDS separates data contained in frames and passes that to the UI for display.  However, other data in the stream (falling outside of valid frames) is discarded.  For nominal operations this is fine, however; there can be several cases where the data is necessary.

1. Multiplexed downlink: some projects push several types of data to the same output stream. For example, console log messages might be printed between valid frames
2. Debugging: when debugging framing and other issues, it is essential to see the junk data.
3. Data loss prevention: even mangled packets may be useful (and recovered) if certain validation steps are ignored (i.e. checksuming).  Discarding the data prevents any recovery.

The GDS should add an option to log this data in the case that it is necessary.

1. "-" can log to standard out.
2. "filename.log" would log to "<log-dir>/filename.log"
3. When nothing is supplied, don't log (current behavior).
4. When just the flag is supplied, create "<log-dir>/unframed.log"

## Rationale

See above.

### PR DESCRIPTION
**Pull request by LeStarch**
| | |
|:---|:---|
|**_Originating Project/Creator_**| Infra |
|**_Affected Component_**|  |
|**_Affected Architectures(s)_**|  |
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  |
|**_Builds Without Errors (y/n)_**|  |
|**_Unit Tests Pass (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

This introduces a new argument `--output-unframed-data` to the `fprime-gds` that allows capture and output of the unframed (junk) data that is discarded by the deframing process.  This is needed for several reasons:

1. Sometimes data is multi-plexed, that is, frames come down mixed with something else (like console data)
2. Sometimes this data is necessary to understand bugs and other issues
3. Sometimes no data loss is acceptable even-if corrupted

This option allows users to output it when the flag is supplied.  When not supplied, the data is discarded as it was previously.

Fixed some other issues:
1. Logs drop-down jumping about: https://github.com/nasa/fprime/issues/2381
2. Logs pane clearing current data when erred preventing user from seeing what *was* there: https://github.com/nasa/fprime/issues/2380
3. Better scrolling of log pane data. Fixes: https://github.com/nasa/fprime/issues/2379
4. Python log output in GDS is no-longer in triplicate. Fixes: https://github.com/nasa/fprime/issues/2378

Primarily fixes: https://github.com/nasa/fprime/issues/2382

## Rationale

See above.

## Testing/Review Recommendations

Tested by putting out-of-frame data into `Ref`'s downlink.

## Future Work

None

### CODE REVIEW

File: src/fprime_gds/executables/comm.py [hunk 1]
```diff
@@ -84,34 +86,59 @@
...
+    discarded_file_handle = None
+    try:
+        if args.output_unframed_data == "-":
+            discarded_file_handle = sys.stdout.buffer
+        elif args.output_unframed_data is not None:
+            discarded_file_handle_path = (
+                Path(args.logs) / Path(args.output_unframed_data)
+            ).resolve()
+            try:
+                discarded_file_handle = open(discarded_file_handle_path, "wb")
```
[the remark this answers is not in the document]
Comment (LeStarch) [reply src/fprime_gds/executables/comm.py:103]: It is always closed when it is opened.  It is only not closed when:

1. Not supplied (None, unused)
2. Is stdout, opened and closed py Python

File: src/fprime_gds/flask/static/js/vue-support/log.js [hunk 2]
```diff
@@ -21,16 +21,19 @@ let template = `
...
+            <input name="scroll" type="checkbox" v-model="scroll" />
+            <label for="scroll">Scroll Log Output</label>
         </div>
     </div>
     <div class="fp-scroll-container">
         <div class="fp-scrollable fp-color-logging">
             <pre><code>{{ text }}</code></pre>
         </div>
     </div>
+    <span>{{ error }}</span>
```
Comment (thomas-bc) [inline src/fprime_gds/flask/static/js/vue-support/log.js]: If you're making changes before merging, could be worth having this be bold and/or red-ish to make it stand out a little

**Review by thomas-bc — DISMISSED** [review]
This is awesome! Looks good to me

**Comment by LeStarch** [comment]
Fixed.  See screenshot:
error

**Review by thomas-bc — APPROVED** [review]

**Pull request merged**